examples/adaptive_crag/application.py

```diff
@@ -0,0 +1,537 @@
+            router.bind(attempts=ATTEMPTS),
+            rewrite_query_for_lancedb.bind(rewrite_attempts=ATTEMPTS),
+            search_lancedb,
+            remove_irrelevant_lancedb_results,
+            extract_keywords_for_exa_search.bind(extraction_attempts=ATTEMPTS),
+            search_exa,
+            ask_assistant.bind(attempts=ATTEMPTS),
+            terminate,
+        )
+        .with_transitions(
+            # if the user wants to exit the conversation
+            ("router", "terminate", when(route="terminate")),  # type: ignore
+            # if the `query` can be directly answered by the assistant
+            ("router", "ask_assistant", when(route="assistant")),  # type: ignore
+            # if a web search is needed
+            ("router", "extract_keywords_for_exa_search", when(route="web_search")),  # type: ignore
+            # the database can have multiple tables for different topics
+            # so it makes sense to have `when` conditions for all other routes and consider this as the default route
+            # otherwise we would have to add a `when` condition for each table name
+            ("router", "rewrite_query_for_lancedb"),
+            ("rewrite_query_for_lancedb", "search_lancedb"),
+            ("search_lancedb", "remove_irrelevant_lancedb_results"),
+            # if there aren't enough relevant documents left after filtering
+            # use web search to find more information
+            (
+                "remove_irrelevant_lancedb_results",
+                "extract_keywords_for_exa_search",
+                expr("len(lancedb_results) < docs_limit"),  # type: ignore
+            ),
+            # if there are enough relevant documents even after filtering, ask the assistant for an answer
+            ("remove_irrelevant_lancedb_results", "ask_assistant"),
+            ("extract_keywords_for_exa_search", "search_exa"),
+            ("search_exa", "ask_assistant"),
+            # go back to the router every time the assistant is done answering
+            ("ask_assistant", "router"),
+        )
+        .with_tracker("local", project=project)
+        .with_identifiers(app_id=app_id, partition_key=username)  # type: ignore
+        .initialize_from(
+            tracker,
+            resume_at_next_action=True,
+            default_entrypoint="router",
+            default_state=dict(db=db, chat_history=[], docs_limit=DOCS_LIMIT),
+        )
+    )
+    return builder.build()
+
+
+if __name__ == "__main__":
+    lance_model = (
+        get_registry()
+        .get("sentence-transformers")
+        .create(name=EMS_MODEL, device=DEVICE)
+    )
+
+    # define the schema of your data using Pydantic
+    class LanceDoc(LanceModel):
+        text: str = lance_model.SourceField()
+        vector: Vector(dim=lance_model.ndims()) = lance_model.VectorField()  # type: ignore
+        file_name: str
+
+    lance_db = lancedb.connect(LANCE_URI)
+
+    burr_docs = load_burr_docs(burr_docs_dir=BURR_DOCS_DIR)
```

Comment:
We should expose this separately as a function with a small CLI. Then put it in the README.

Note it needs to be done, in advance, to get it to work. Maybe add a check in the Burr workflow to see if it exists?

Reply:
How do you want this flow to look?